### PR TITLE
Deprecate DeploymentConfig

### PR DESCRIPTION
**Pull request by edeandrea**
OpenShift's `DeploymentConfig` object has been deprecated as of OpenShift 4.14. See https://access.redhat.com/articles/7041372 for details.

This PR adds a deprecation warning & additional documentation to the config item as well as sets the default value of `quarkus.openshift.deployment-kind` to `deployment` instead of `deployment-config` when `flavor != v3`. For `flavor == v3`, the default value `deployment-config` remains.

This should probably be mentioned in the migration guide as well.

I updated existing tests and added a handful of new ones to verify old and new behavior.

### CODE REVIEW

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/OpenshiftConfig.java [hunk 1]
```diff
@@ -41,6 +41,7 @@ public static enum OpenshiftFlavor {
 
     public static enum DeploymentResourceKind {
         Deployment(DEPLOYMENT, DEPLOYMENT_GROUP, DEPLOYMENT_VERSION),
+        @Deprecated(since = "OpenShift 4.14", forRemoval = true)
```
Comment (Ladicek) [inline extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/OpenshiftConfig.java:44]: The article you link to says

> Also while customers using DeploymentConfigs are encouraged to migrate to Deployments, there are currently (Red Hat OpenShift Container Platform 4.14) no plans to remove the DeploymentConfigs API. Instead the DeploymentConfigs API is being deprecated and therefore no new functionality will be provided and maintenance will be limited to security and critical fixes only.

so `forRemoval = true` seems incorrect?

Comment (edeandrea) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/OpenshiftConfig.java:44]: You are probably right. I will fix that.

Comment (edeandrea) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/OpenshiftConfig.java]: @Ladicek I've fixed this.

File: docs/src/main/asciidoc/deploying-to-openshift.adoc [hunk 2]
```diff
@@ -96,8 +96,19 @@ TIP: If you want to test your application immediately then set the `quarkus.open
...
 
 This command will build your application locally, then trigger a container image build and finally apply the generated OpenShift resources automatically.
-The generated resources use OpenShift's `DeploymentConfig` that is configured to automatically trigger a redeployment when a change in the `ImageStream` is noticed.
-In other words, any container image build after the initial deployment will automatically trigger redeployment, without the need to delete, update or re-apply the generated resources.
+The generated resources use a Kubernetes `Deployment`, but still make use of OpenShift specific resources like `Route`, `BuildConfig` etc.
+
+IMPORTANT: As of OpenShift 4.14, the https://access.redhat.com/articles/7041372[`DeploymentConfig` object has been deprecated].
+
+Since `Deployment` is a Kubernetes resource and not OpenShift specific, it can't possibly leverage `ImageStream` resources, as is the case with `DeploymentConfig`. This means that the image references need to include the container image registry that hosts the image.
```
Comment (iocanel) [inline docs/src/main/asciidoc/deploying-to-openshift.adoc:103]: The container image registry can be omitted, when the local lookup policy is enabled on the `ImageStream`. This policy is enabled for all image streams created by the extension.

**Review by iocanel — CHANGES_REQUESTED** [review]
Thank you so much for this. Looks great!
I've requested a minor change in the added doc.

File: docs/src/main/asciidoc/deploying-to-openshift.adoc (hunk 2 above)
Comment (iocanel) [reply docs/src/main/asciidoc/deploying-to-openshift.adoc:103]: Ok, then let's address that in an other PR.

**Review by iocanel — APPROVED** [review]
LGTM

**Comment by edeandrea** [comment]
Any news on when we can get this merged?

**Comment by yrodiere** [comment]
> This should probably be mentioned in the migration guide as well.

@edeandrea Yeah, I think it should... I just got bitten by this, ended up with both a `DeploymentConfig` (the old one) and a `Deployment` (the new one) in my prod env... I can tell you it's not great.

Can someone please document this (and other breaking changes, if any) in the migration guide of 3.7?

**Comment by edeandrea** [comment]
I don't have access to the migration guide/wiki. The documentation in the guides has been updated accordingly.

**Comment by edeandrea** [comment]
@gsmet / @yrodiere This should be mentioned in the [3.7 migration guide](https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.7).

I don't have permissions to edit the wiki. I'd be happy to add something about it if I did.

**Comment by yrodiere** [comment]
> @gsmet / @yrodiere This should be mentioned in the [3.7 migration guide](https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.7).
> 
> I don't have permissions to edit the wiki. I'd be happy to add something about it if I did.

@maxandersen probably should give you the appropriate permissions.

**Pull request merged**